### PR TITLE
[FSDP2] Fix memory spike with `cpu_ram_efficient_loading=True`

### PR DESCRIPTION
**Pull request by S1ro1**
### With initial implementation of `FSDP2`, there is an issue if `cpu_ram_efficient_loading` is set. See [this comment](https://github.com/huggingface/accelerate/blob/23b707bc0c290c8b727d54ae4c487283f969265f/src/accelerate/utils/fsdp_utils.py#L584-L587) for details. This usually isn't an issue, but if the model is large enough, it can lead to memory spike that is ~2x the size of peak.

This PR fixes the spike first found in #3474. Providing some numerics below:
Loss parity between `cpu_ram_efficient_loading=true/false`
[wandb link](https://wandb.ai/siro/PR3482?nw=nwusersiro)

Memory snapshots:
### Old (`cpu_ram_efficient_loading=True`)

image

### New (`cpu_ram_efficient_loading=True`)

image

### Baseline (`cpu_ram_efficient_loading=False`)

image

(The discrepancy between new/baseline in the semi-big rectangle appearing first/in the middle is caused by difference in order of evaluation - top-level first vs bottom-level first)

cc @SunMarc @zach-huggingface 

Possibly moving `get_non_persistent_buffer_fqns` and making it `private` could be an upgrade, this function is quite specific to this use-case.

### CODE REVIEW

File: src/accelerate/utils/fsdp_utils.py [hunk 1]
```diff
@@ -545,6 +580,26 @@ def fsdp2_prepare_model(accelerator, model: torch.nn.Module) -> torch.nn.Module:
...
+        # Context: `fully_shard` moves the model to GPU if it was on CPU, however it can also be on `meta` and then it stays there even after `fully_shard`
+        # For this reason, we need to move the model to `meta` device, as then sharding happens on `meta` device
+        # If we kept the model on CPU (`cpu_ram_efficient_loading` has model be on CPU on all ranks, though non-main ranks only have `torch.emtpy`), `fully_shard` would move it to GPU
+        # Afterwards, when we call `fsdp2_load_full_state_dict`, us creating the state_dict would result into briefly having two copies of model state_dict on the GPU -> VRAM spike
+
+        # We need to keep the original buffers, as those MAY not be in the state_dict, resulting in them staying on meta device
+        # Also, these buffers aren't getting sharded by default
+        original_named_buffers = copy.deepcopy(dict(model.named_buffers()))
+        # We get the FQNs of all non-persistent buffers, to re-register them after
+        non_persistent_buffer_fqns = get_non_persistent_buffer_fqns(model)
```
Comment (SunMarc) [inline src/accelerate/utils/fsdp_utils.py:593]: only the non persistant buffers are not in the `state_dict`, so let's not copy all buffers

**Review by SunMarc — COMMENTED** [review]
Thanks for the PR ! left a couple of comments

File: src/accelerate/utils/fsdp_utils.py [hunk 2]
```diff
@@ -434,29 +435,37 @@ def fsdp2_load_full_state_dict(accelerator, model: torch.nn.Module, full_sd: dic
...
             full_tensor = torch.empty(sharded_param.size(), device="cuda", dtype=sharded_param.dtype)
             mesh = sharded_param.device_mesh
             dist.broadcast(full_tensor, src=0, group=mesh.get_group())
             sharded_tensor = distribute_tensor(full_tensor, mesh, sharded_param.placements)
             sharded_sd[param_name] = sharded_tensor
 
-    model.load_state_dict(sharded_sd)
+    # we set `assign=True` because our params are on meta device
+    model.load_state_dict(sharded_sd, assign=True)
+    return model
```
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/accelerate/utils/fsdp_utils.py:499]: Is this really required for this use-case? This is to be only used internally with `cpu_ram_efficient_loading` where the `meta_sd` already has gone over the `infer_dtype` logic. Therefore no need to fix it again.

[the remark this answers is not in the document]
Comment (SunMarc) [reply src/accelerate/utils/fsdp_utils.py:499]: Maybe not, i'm not sure at all. If you can try if a model that where the state dict is in bf16 but want to load the model in fp32 (torch_dtype=fp32 in `from_pretrained`), that could help figuring out.

File: src/accelerate/utils/fsdp_utils.py [hunk 3]
```diff
@@ -434,29 +435,68 @@ def fsdp2_load_full_state_dict(accelerator, model: torch.nn.Module, full_sd: dic
...
             full_param = full_param.detach().cuda()
             mesh = sharded_param.device_mesh
             dist.broadcast(full_param, src=0, group=mesh.get_group())
             sharded_tensor = distribute_tensor(full_param, mesh, sharded_param.placements)
+            to_contiguous, casting_dtype = _infer_parameter_dtype(
+                model,
+                param_name,
+                full_param,
+            )
+            sharded_tensor = _cast_and_contiguous(sharded_tensor, to_contiguous, casting_dtype)
```
Comment (SunMarc) [inline src/accelerate/utils/fsdp_utils.py:480]: can't we directly use transformers private method as I think this is a special case for transformers model ?

File: src/accelerate/utils/fsdp_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/accelerate/utils/fsdp_utils.py:499]: but otherwise, if you think this is not a big deal, we can also fix it later or not

File: src/accelerate/utils/fsdp_utils.py (hunk 3 above)
Comment (S1ro1) [reply src/accelerate/utils/fsdp_utils.py:480]: I wanted to avoid relying on it, the transformers method introduces some extra functionality and this is simple enough that we can keep it in here. It's a bit more code, but more readable IMO and makes us not rely on transformers for this.

File: src/accelerate/utils/fsdp_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/accelerate/utils/fsdp_utils.py:499]: nice !

**Review by SunMarc — APPROVED** [review]
Thanks a lot for iterating ! Feel free to merge it
